### PR TITLE
ddl:support drop index for the partitioned table.

### PR DESCRIPTION
**Pull request by ciscoxll**
## What have you changed? (mandatory)
support drop index for the partitioned table
* We first need to get all the partition IDs and then we need to delete the index in each partition.
* Fix issue #7222 

Take this as an example:
```sql
mysql> set @@tidb_enable_table_partition = 1;
Query OK, 0 rows affected (0.00 sec)

mysql> create table tr(
    ->         id int, name varchar(50),
    ->         purchased date
    ->     )
    ->     partition by range( year(purchased) ) (
    ->         partition p0 values less than (1990),
    ->         partition p1 values less than (1995),
    ->         partition p2 values less than (2000),
    ->         partition p3 values less than (2005),
    ->         partition p4 values less than (2010),
    ->         partition p5 values less than (2015)
    ->        );
Query OK, 0 rows affected (0.01 sec)

mysql>  INSERT INTO tr VALUES
    ->     (1, 'desk organiser', '2003-10-15'),
    ->     (2, 'alarm clock', '1997-11-05'),
    ->     (3, 'chair', '2009-03-10'),
    ->     (4, 'bookcase', '1989-01-10'),
    ->     (5, 'exercise bike', '2014-05-09'),
    ->     (6, 'sofa', '1987-06-05'),
    ->     (7, 'espresso maker', '2011-11-22'),
    ->     (8, 'aquarium', '1992-08-04'),
    ->     (9, 'study desk', '2006-09-16'),
    ->     (10, 'lava lamp', '1998-12-25');
Query OK, 10 rows affected (0.10 sec)

mysql> create index idx1 on tr (purchased);
Query OK, 0 rows affected (0.04 sec)

mysql> select count(purchased) from tr use index(idx1);
+------------------+
| count(purchased) |
+------------------+
|               10 |
+------------------+
1 row in set (0.00 sec)

mysql> drop index idx1 on tr;
Query OK, 0 rows affected (0.03 sec)

mysql> select count(purchased) from tr use index(idx1);
ERROR 1176 (42000): Key 'idx1' doesn't exist in table 'tr'
```

## What is the type of the changes? (mandatory)

New feature
- New feature (non-breaking change which adds functionality)

## How has this PR been tested? (mandatory)
Unit tests.

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

## Does this PR need to be added to the release notes? (mandatory)
No

### CODE REVIEW

File: ddl/db_test.go [hunk 1]
```diff
@@ -3096,6 +3096,59 @@ func (s *testDBSuite) TestPartitionUniqueKeyNeedAllFieldsInPf(c *C) {
...
+    (3, 'chair', '2009-03-10'),
+    (4, 'bookcase', '1989-01-10'),
+    (5, 'exercise bike', '2014-05-09'),
+    (6, 'sofa', '1987-06-05'),
+    (7, 'espresso maker', '2011-11-22'),
+    (8, 'aquarium', '1992-08-04'),
+    (9, 'study desk', '2006-09-16'),
+    (10, 'lava lamp', '1998-12-25');`)
+
+	tk.MustExec("create index idx1 on tr (purchased)")
```
Comment (tiancaiamao) [inline ddl/db_test.go]: It test drop index, so how about create the table with index together?

File: ddl/db_test.go [hunk 2]
```diff
@@ -3096,6 +3096,59 @@ func (s *testDBSuite) TestPartitionUniqueKeyNeedAllFieldsInPf(c *C) {
...
+	tk.MustQuery("select count(purchased) from tr use index(idx1)").Check(testkit.Rows("10"))
+	tk.MustExec("admin check table tr")
+	// Wait for add index done.
+	time.Sleep(waitForCleanDataInterval)
```
Comment (tiancaiamao) [inline ddl/db_test.go]: There is no need to wait.
 `create index idx1 on tr (purchased)`  blocks for a while, as long as it returns, add index will be done.

File: ddl/index.go [hunk 3]
```diff
@@ -436,7 +436,7 @@ func onDropIndex(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 			job.Args[0] = indexInfo.ID
 		} else {
 			job.FinishTableJob(model.JobStateDone, model.StateNone, ver, tblInfo)
-			job.Args = append(job.Args, indexInfo.ID)
+			job.Args = append(job.Args, indexInfo.ID, getPartitionIDs(tblInfo))
```
Comment (tiancaiamao) [inline ddl/index.go]: If the table is not a partition, what will `getPartitionIDs` return?

Comment (ciscoxll) [reply ddl/index.go]: `getPartitionIDs` will return `[]int64{}`.

File: table/tables/tables.go [hunk 4]
```diff
@@ -81,10 +81,10 @@ func MockTableFromMeta(tblInfo *model.TableInfo) table.Table {
 
 	var t Table
 	initTableCommon(&t.tableCommon, tblInfo, tblInfo.ID, columns, nil)
+	if err := initTableIndices(&t.tableCommon); err != nil {
```
Comment (tiancaiamao) [inline table/tables/tables.go]: Is this change relevant to the PR?

**Comment by ciscoxll** [comment]
@tiancaiamao @winkyao PTAL.

**Comment by tiancaiamao** [comment]
LGTM

**Comment by ciscoxll** [comment]
@winkyao @zimulala PTAL.

File: ddl/delete_range.go [hunk 5]
```diff
@@ -279,9 +279,20 @@ func insertJobIntoDeleteRangeTable(ctx sessionctx.Context, job *model.Job) error
 		tableID := job.TableID
 		var indexName interface{}
 		var indexID int64
-		if err := job.DecodeArgs(&indexName, &indexID); err != nil {
+		var partitionIDs []int64
+		if err := job.DecodeArgs(&indexName, &indexID, &partitionIDs); err != nil {
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply ddl/delete_range.go]: When modifying table schema, there is a new table info and an old one.
It's easy to get the stale IDs if we get it from the info schema. @shenli

[the remark this answers is not in the document]
Comment (ciscoxll) [reply ddl/delete_range.go]: @shenli Existing function `getPartitionIDs`.
```go
func getPartitionIDs(table *model.TableInfo) []int64 {
	if table.GetPartitionInfo() == nil {
		return []int64{}
	}
	partitionIDs := make([]int64, 0, len(table.Partition.Definitions))
	for _, def := range table.Partition.Definitions {
		partitionIDs = append(partitionIDs, def.ID)
	}
	return partitionIDs
}
```
Is it clearer.

[the remark this answers is not in the document]
Comment (crazycs520) [reply ddl/delete_range.go]: In `drop index`, because the `ddl` of same table is serial executed, so get partition IDs from info schema is ok.
But I remain neutral, get the partition IDs from the info schema or use job arguments are both ok for me. >v<

[the remark this answers is not in the document]
Comment (shenli) [reply ddl/delete_range.go]: "It's easy to get the stale IDs if we get it from the info schema." Can not understand this. We should guarantee that we will get the correct schema when doing DDL.

[the remark this answers is not in the document]
Comment (tiancaiamao) [reply ddl/delete_range.go]: You can ignore the comment @shenli , I just give a hint on how to achieve this...

> We should guarantee that we will get the correct schema when doing DDL.

[the remark this answers is not in the document]
Comment (shenli) [reply ddl/delete_range.go]: @ciscoxll Please address the comment. We do not need to put partition IDs in the job argument.

**Comment by ciscoxll** [comment]
@winkyao @zimulala PTAL.

**Comment by ciscoxll** [comment]
@shenli @winkyao @crazycs520 PTAL.

File: ddl/delete_range.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (zimulala) [reply ddl/delete_range.go]: @shenli @winkyao 
I think it's OK. The processing here is unified with other places.
